Fix Windows Build

ZacharyYarostTechsmith opened the pull request:
Card: https://github.com/Samsung/rlottie/issues/576

Hello, thank you very much for the recent vulnerability patch! While doing updates we noticed that rlottie does not currently build on windows with the changes.

A recent commit: https://github.com/Samsung/rlottie/commit/346c9b604b4227b491f38e1c5a4dd43d4f7c45ac from mid-June uses PATH_MAX which isn't defined on Windows.

Additionally, `realpath` is a POSIX function: https://man7.org/linux/man-pages/man3/realpath.3.html so an equivalent is needed for Windows.

This PR resolves those two issues.

ajorians committed:
Added an anonymous namespace method to canonize a path.

I also needed to define PATH_MAX which needed include from Windows.h.  I included shlwapi.h for PathCanonicalizeA.

ajorians committed:
Use PathCanonicalizeW on Windows.

I first converted to std::wstring; and then back after conversion.  Added library to link with.

ZacharyYarostTechsmith commented:
@gindemit Thank you for the approval! Could I ask you to merge this PR since I do not have permissions?

gindemit commented:
Hi @ZacharyYarostTechsmith, I just reviewed your changes and approved those. But I don't have the permissions as well.

ZacharyYarostTechsmith commented:
Ah, I understand. Thank you for making the review none the less.

ZacharyYarostTechsmith commented:
@JoogabYun Would you be open to merging this update?

I see that you've merged the previous related PRs.

laggykiller commented:
https://github.com/TechSmith/rlottie/blob/d89b8f1ca33ee476fc0a397a0a24f516dafd42c6/src/lottie/lottieparser.cpp#L814
```c++
   std::wstring ToStdWString( std::string_view str )
```

This will not compile below C++17. I suggest `std::wstring ToStdWString(const std::string &str)`

---

https://github.com/TechSmith/rlottie/blob/d89b8f1ca33ee476fc0a397a0a24f516dafd42c6/src/lottie/lottieparser.cpp#L829
```c++
   std::string ToStdString( std::wstring_view wstr )
```

This will not compile below C++17. I suggest `std::string ToStdString(const std::wstring &wstr)`

---

https://github.com/TechSmith/rlottie/blob/d89b8f1ca33ee476fc0a397a0a24f516dafd42c6/src/lottie/lottieparser.cpp#L850

```c++
       if ( PathCanonicalizeW( wresolved_path.data(), wpath.c_str() ) )
```

This will not compile in older standards and may throw `error C2664: 'BOOL PathCanonicalizeW(LPWSTR,LPCWSTR)': cannot convert argument 1 from 'const _Elem *' to 'LPWSTR'`. I suggest `if ( PathCanonicalizeW(const_cast<wchar_t*>(wresolved_path.data()), wpath.c_str() ) )`

ZacharyYarostTechsmith commented:
@laggykiller Thanks for the feedback. Those recommendations are all reasonable. I will make the update when I have a moment to build the changes and try them.

gindemit commented:
btw @ZacharyYarostTechsmith, I used at the beginning your changes and then made some changes on top of those with AI help to fix the windows build for the Unity rlottie plugin. The windows native libraries are working fine for me. But I switched to C++ 17 version in the build process.
You can review the changes [here.](https://github.com/Samsung/rlottie/compare/master...gindemit:rlottie:unity-rlottie-win-fixes)

JoogabYun commented:
@mmaciola 
Please review the patch

mmaciola commented:
Hi @ZacharyYarostTechsmith. Thank you for your contribution! The idea behind this patch is good. Project is currently using cpp_std=c++14. Please apply the suggestions provided by @laggykiller — once that’s done, we’ll be happy to merge it into the project.

ZacharyYarostTechsmith committed:
fixes support for pre C++ 17
- drop string_view params
- use const std::(w)string&
- const_cast PathCanonicalizeW buffer
- remove string_view include

ZacharyYarostTechsmith commented:
@mmaciola The suggestions provided by @laggykiller have been added to this branch and is ready to merge.

@gindemit I'll take a look at your changes now.
